snuba/utils/metrics/__init__.py

```diff
@@ -0,0 +1,59 @@
+from random import random
+from collections import ChainMap
+from typing import Mapping, Optional, Union
+
+
+Tags = Mapping[str, str]
+
+
+from snuba.utils.metrics.backends.abstract import MetricsBackend
+
+
+class Metrics:
+    def __init__(
+        self,
+        backend: MetricsBackend,
+        namespace: Optional[str] = None,
+        tags: Optional[Tags] = None,
+        sample_rates: Optional[Mapping[str, float]] = None,
+    ) -> None:
+        self.__backend = backend
+        self.__namespace = namespace
+        self.__tags = tags
+        self.__sample_rates = sample_rates if sample_rates is not None else {}
+
+    def __apply_namespace(self, name: str) -> str:
+        if self.__namespace is not None:
+            return f"{self.__namespace}.{name}"
+        else:
+            return name
+
+    def __merge_tags(self, tags: Optional[Tags]) -> Tags:
+        return ChainMap(*[d for d in [self.__tags, tags] if d is not None])
+
+    def __is_sampled(self, name: str) -> bool:
```

Comment:
What is the reason for managing the sampling rate in this class instead of passing a sample rate to the backend as we did in cdc ?
I think there is a value in letting datadog perform the sampling: when incrementing the metric it should account for the sampling rate thus incrementing by the expected value (at least I hope it does).
If not should we multiply the value we increment by 1/sampling_rate to have a more reliable number in the dashboard?

Reply:
This was mainly because it seemed like for extensibility's sake, it was easy enough to perform this in the wrapper class rather than deferring it to the backend. I don't have a very strong opinion about that, though — just a preference. It does make it easier to test that the sampling rate is being correctly (for some definition of correctly) applied without mocking, too.

I'm not sure that Datadog itself (or statsd) adjusts values based on the sampling rate. I'll take a look so that we can make a more informed decision about what to do here.

Reply:
It looks like Datadog performs sampling the same way:

https://github.com/DataDog/datadogpy/blob/b4cb2df6704ebdcf9fcd0d72b2251da5111b41cf/datadog/dogstatsd/base.py#L305-L306

… _but_, the sampling rate is also part of the UDP payload, so I think it'd be helpful for the Datadog backend to have knowledge of these rates in case it does anything useful with them:

https://github.com/DataDog/datadogpy/blob/b4cb2df6704ebdcf9fcd0d72b2251da5111b41cf/datadog/dogstatsd/base.py#L311-L319

If we pull the `sample_rate` mapping [from here](https://github.com/getsentry/snuba/blob/32628d2fb60cfd0b258e342342bcb260654636fa/snuba/utils/metrics/__init__.py#L18) and move it to the Datadog backend, that also means that the namespacing [from here](https://github.com/getsentry/snuba/blob/32628d2fb60cfd0b258e342342bcb260654636fa/snuba/utils/metrics/__init__.py#L16) would also have to be part of the backend. This makes the only thing left at the higher level of abstraction the constant tags, which seems sort of like a dubious value proposition at that point, considering the Datadog metrics backend will also do that as well. I think in that case, it probably just makes sense to instantiate the `MetricsBackend` classes directly instead.

Any thoughts before I make that change.

Reply:
About what datadog does with the sampling rate. It is true, it locally processes it like this PR was doing, but as you mentioned, the critical point is passing the rate to datadog so that it can take that into account when displaying data:
https://docs.datadoghq.com/developers/faq/dog-statsd-sample-rate-parameter-explained/

Whether we should keep a higher level abstraction only for the constant tags or not, if we can just instantiate the backend and use that, that makes things simpler and it would not be hard to introduce an additional level on top of it if needed so I like the idea of going directly to the backend. Not a strong opinion though since the abstraction is very simple anyway.

Reply:
I went ahead and just made that chance because I figured at worst, it's an easy revert: 69eda4e5e585a87d3b0bb5968ecbdb5f84d75760

I'm generally OK with this, the only thing I don't particularly like is this mismatch that I introduced by making tags a mapping instead of a list:

https://github.com/getsentry/snuba/blob/69eda4e5e585a87d3b0bb5968ecbdb5f84d75760/snuba/util.py#L611